# Random Forest based on selected outputs 
# Preprocessing of csv file 

# preprocessing smaller dataset for now 

In [64]:
from pathlib import Path
import pandas as pd

# Location of the Elliptic++ dataset
DATA_DIR = Path("../Elliptic++ Dataset")
TX_FEATURES = DATA_DIR / "txs_features.csv"
# -----------------------------------------
# 1. Get the column names
# -----------------------------------------

# nrows=0 means we only read the header,
# not the entire CSV
columns = pd.read_csv(
    TX_FEATURES,
    nrows=0
).columns.tolist()

print("Number of columns:", len(columns))

# Print every column and its position
for i, column in enumerate(columns):
    print(i, column)


# -----------------------------------------
# 2. Select txId and the last 10 columns
# -----------------------------------------

# First column contains the transaction ID
tx_id = columns[0]

# Select the last 10 feature columns
last_10_columns = columns[-10:]

# Keep txId along with those 10 features
columns_to_keep = [tx_id] + last_10_columns

print("\nTransaction ID:")
print(tx_id)

print("\nLast 10 columns:")

for column in last_10_columns:
    print(column)


# -----------------------------------------
# 3. Load only the columns we need
# -----------------------------------------

tx_data = pd.read_csv(
    TX_FEATURES,
    usecols=columns_to_keep
)

display(tx_data.head())


# -----------------------------------------
# 4. Check the preprocessed data
# -----------------------------------------

print("Shape:", tx_data.shape)

print("\nColumns:")
print(tx_data.columns.tolist())

print("\nData types:")
print(tx_data.dtypes)

# -----------------------------------------
# 5. Check for missing values
# -----------------------------------------

print("\nMissing values:")
print(tx_data.isnull().sum())


# -----------------------------------------
# 6. Check for duplicate transaction IDs
# -----------------------------------------

print("\nDuplicate transaction IDs:")
print(tx_data[tx_id].duplicated().sum())

Number of columns: 184
0 txId
1 Time step
2 Local_feature_1
3 Local_feature_2
4 Local_feature_3
5 Local_feature_4
6 Local_feature_5
7 Local_feature_6
8 Local_feature_7
9 Local_feature_8
10 Local_feature_9
11 Local_feature_10
12 Local_feature_11
13 Local_feature_12
14 Local_feature_13
15 Local_feature_14
16 Local_feature_15
17 Local_feature_16
18 Local_feature_17
19 Local_feature_18
20 Local_feature_19
21 Local_feature_20
22 Local_feature_21
23 Local_feature_22
24 Local_feature_23
25 Local_feature_24
26 Local_feature_25
27 Local_feature_26
28 Local_feature_27
29 Local_feature_28
30 Local_feature_29
31 Local_feature_30
32 Local_feature_31
33 Local_feature_32
34 Local_feature_33
35 Local_feature_34
36 Local_feature_35
37 Local_feature_36
38 Local_feature_37
39 Local_feature_38
40 Local_feature_39
41 Local_feature_40
42 Local_feature_41
43 Local_feature_42
44 Local_feature_43
45 Local_feature_44
46 Local_feature_45
47 Local_feature_46
48 Local_feature_47
49 Local_feature_48
50 Local_featur

,txId,in_BTC_min,in_BTC_max,in_BTC_mean,in_BTC_median,in_BTC_total,out_BTC_min,out_BTC_max,out_BTC_mean,out_BTC_median,out_BTC_total
0,3321,0.534072,0.534072,0.534072,0.534072,0.534072,1.668990e-01,0.367074,0.266986,0.266986,0.533972
1,11108,5.611878,5.611878,5.611878,5.611878,5.611878,5.861940e-01,5.025584,2.805889,2.805889,5.611778
2,51816,0.456608,0.456608,0.456608,0.456608,0.456608,2.279902e-01,0.228518,0.228254,0.228254,0.456508
3,68869,0.308900,8.000000,3.102967,1.000000,9.308900,1.229000e+00,8.079800,4.654400,4.654400,9.308800
4,89273,852.164680,852.164680,852.164680,852.164680,852.164680,1.300000e-07,41.264036,0.065016,0.000441,852.164680


Shape: (203769, 11)

Columns:
['txId', 'in_BTC_min', 'in_BTC_max', 'in_BTC_mean', 'in_BTC_median', 'in_BTC_total', 'out_BTC_min', 'out_BTC_max', 'out_BTC_mean', 'out_BTC_median', 'out_BTC_total']

Data types:
txId                int64
in_BTC_min        float64
in_BTC_max        float64
in_BTC_mean       float64
in_BTC_median     float64
in_BTC_total      float64
out_BTC_min       float64
out_BTC_max       float64
out_BTC_mean      float64
out_BTC_median    float64
out_BTC_total     float64
dtype: object

Missing values:
txId                0
in_BTC_min        965
in_BTC_max        965
in_BTC_mean       965
in_BTC_median     965
in_BTC_total      965
out_BTC_min       965
out_BTC_max       965
out_BTC_mean      965
out_BTC_median    965
out_BTC_total     965
dtype: int64

Duplicate transaction IDs:
0


# removed the 965 rows in the dataset instead of filling with zeros found after tests 

In [71]:
# Remove rows where the BTC features are missing

print("Rows before cleaning:", len(tx_data))

tx_data_clean = tx_data.dropna(
    subset=last_10_columns
).copy()

print("Rows after cleaning:", len(tx_data_clean))
print("Rows removed:", len(tx_data) - len(tx_data_clean))

print("\nMissing values after cleaning:")
print(tx_data_clean.isnull().sum())

Rows before cleaning: 203769
Rows after cleaning: 202804
Rows removed: 965

Missing values after cleaning:
txId              0
in_BTC_min        0
in_BTC_max        0
in_BTC_mean       0
in_BTC_median     0
in_BTC_total      0
out_BTC_min       0
out_BTC_max       0
out_BTC_mean      0
out_BTC_median    0
out_BTC_total     0
dtype: int64


In [78]:
print("PREPROCESSING TESTS")

# Make sure the dataset isn't empty
assert not tx_data_clean.empty

# Should have txId + 10 features
assert tx_data_clean.shape[1] == 11

# Make sure txId is still there
assert tx_id in tx_data_clean.columns

# Make sure there are no duplicate txIds
assert tx_data_clean[tx_id].duplicated().sum() == 0

# Make sure there are no missing values
assert tx_data_clean[last_10_columns].isnull().sum().sum() == 0

print("Duplicate transaction IDs:",
      tx_data_clean[tx_id].duplicated().sum())

print("Missing values:",
      tx_data_clean[last_10_columns].isnull().sum().sum())

print("\nAll preprocessing tests passed!")

display(tx_data_clean.head())

PREPROCESSING TESTS
Duplicate transaction IDs: 0
Missing values: 0

All preprocessing tests passed!


,txId,in_BTC_min,in_BTC_max,in_BTC_mean,in_BTC_median,in_BTC_total,out_BTC_min,out_BTC_max,out_BTC_mean,out_BTC_median,out_BTC_total
0,3321,0.534072,0.534072,0.534072,0.534072,0.534072,1.668990e-01,0.367074,0.266986,0.266986,0.533972
1,11108,5.611878,5.611878,5.611878,5.611878,5.611878,5.861940e-01,5.025584,2.805889,2.805889,5.611778
2,51816,0.456608,0.456608,0.456608,0.456608,0.456608,2.279902e-01,0.228518,0.228254,0.228254,0.456508
3,68869,0.308900,8.000000,3.102967,1.000000,9.308900,1.229000e+00,8.079800,4.654400,4.654400,9.308800
4,89273,852.164680,852.164680,852.164680,852.164680,852.164680,1.300000e-07,41.264036,0.065016,0.000441,852.164680


In [79]:
FEATURE_COLS = [
    # Current Transaction
    "input_count",
    "output_count",
    "input_total",
    "output_total",
    "fee",

    # Input Address History

    # Prior transaction count
    "input_prior_tx_count_24h",
    "input_prior_tx_count_7d",
    "input_prior_tx_count_total",

    # Average prior transaction count
    "input_prior_tx_avg_24h",
    "input_prior_tx_avg_7d",
    "input_prior_tx_avg_total",

    # Maximum previous transaction count
    "input_prev_tx_max_24h",
    "input_prev_tx_max_7d",
    "input_prev_tx_max_total",

    # Bitcoin sent
    "input_btc_sent_24h",
    "input_btc_sent_7d",
    "input_btc_sent_total",

    # Bitcoin received
    "input_btc_received_24h",
    "input_btc_received_7d",
    "input_btc_received_total",

    # Average BTC transaction amount
    "input_btc_avg_amount_24h",
    "input_btc_avg_amount_7d",
    "input_btc_avg_amount_total",

    # Maximum BTC transaction amount
    "input_btc_max_amount_24h",
    "input_btc_max_amount_7d",
    "input_btc_max_amount_total",

    # Output Address History

    "output_prior_tx_count_24h",
    "output_prior_tx_count_7d",
    "output_prior_tx_count_total",

    "output_prior_tx_avg_24h",
    "output_prior_tx_avg_7d",
    "output_prior_tx_avg_total",

    "output_prev_tx_max_24h",
    "output_prev_tx_max_7d",
    "output_prev_tx_max_total",

    "output_btc_sent_24h",
    "output_btc_sent_7d",
    "output_btc_sent_total",

    "output_btc_received_24h",
    "output_btc_received_7d",
    "output_btc_received_total",

    "output_btc_avg_amount_24h",
    "output_btc_avg_amount_7d",
    "output_btc_avg_amount_total",

    "output_btc_max_amount_24h",
    "output_btc_max_amount_7d",
    "output_btc_max_amount_total",

    # Historical Relationship

    "fraction_address_pairs_seen_before"
]

In [80]:
OPT_FEATURE_COLS = [

    "time_since_last_interaction_max",
    "time_since_last_interaction_avg",

    "value_between_pairs_max",
    "value_between_pairs_avg",

    "interaction_count_total",
    "interaction_count_max",
    "interaction_count_avg",

    "bidirectional_relationship_value"
]

In [81]:
#imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay,
    roc_curve, 
    roc_auc_score,
    precision_score,
    recall_score,
    f1_score
)

In [82]:
# need to make sure csv is cleaned **
# 0 = legal 
# 1 = scam 
data = pd.read_csv("transaction_features.csv")

print(data.shape)
display(data.head())

FileNotFoundError: [Errno 2] No such file or directory: 'transaction_features.csv'

In [ ]:
# define and initialize feutures 
feature_cols = [

    # Current transaction
    "input_count",
    "output_count",
    "input_total",
    "output_total",
    "fee",

    # Input address history
    "input_prior_tx_count_24h",
    "input_prior_tx_count_7d",
    "input_prior_tx_count_total",

    "input_prior_tx_avg_24h",
    "input_prior_tx_avg_7d",
    "input_prior_tx_avg_total",

    "input_prev_tx_max_24h",
    "input_prev_tx_max_7d",
    "input_prev_tx_max_total",

    "input_btc_sent_24h",
    "input_btc_sent_7d",
    "input_btc_sent_total",

    "input_btc_received_24h",
    "input_btc_received_7d",
    "input_btc_received_total",

    "input_btc_avg_amount_24h",
    "input_btc_avg_amount_7d",
    "input_btc_avg_amount_total",

    "input_btc_max_amount_24h",
    "input_btc_max_amount_7d",
    "input_btc_max_amount_total",

    # Output address history
    "output_prior_tx_count_24h",
    "output_prior_tx_count_7d",
    "output_prior_tx_count_total",

    "output_prior_tx_avg_24h",
    "output_prior_tx_avg_7d",
    "output_prior_tx_avg_total",

    "output_prev_tx_max_24h",
    "output_prev_tx_max_7d",
    "output_prev_tx_max_total",

    "output_btc_sent_24h",
    "output_btc_sent_7d",
    "output_btc_sent_total",

    "output_btc_received_24h",
    "output_btc_received_7d",
    "output_btc_received_total",

    "output_btc_avg_amount_24h",
    "output_btc_avg_amount_7d",
    "output_btc_avg_amount_total",

    "output_btc_max_amount_24h",
    "output_btc_max_amount_7d",
    "output_btc_max_amount_total",

    # Relationship history
    "fraction_address_pairs_seen_before"
]
X = data[feature_cols]

y = data["class"]

print("Features:", X.shape)
print("Labels:", y.shape)


NameError: name 'data' is not defined

In [ ]:
# check data 
print(X.info())

print("\nMissing values:")
print(X.isnull().sum().sort_values(ascending=False).head(20))

print("\nClass distribution:")
print(y.value_counts())

print("\nClass percentages:")
print(y.value_counts(normalize=True))
X = X.dropna(0)

In [ ]:
# check for impossible/ infinite vals 
data["input_prior_tx_avg_24h"]
X = X.replace([np.inf, -np.inf], np.nan)

X = X.fillna(0)
# print statements to check for both 
print("Missing values:", X.isnull().sum().sum())
print("Infinite values:", np.isinf(X.to_numpy()).sum())

In [ ]:
# Train/ test the split 
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y # for imbalance 
)

print("Training examples:", len(X_train))
print("Testing examples:", len(X_test))

In [ ]:
# Random Forest 
rf_model = RandomForestClassifier(
    n_estimators=300,
    random_state=42,
    class_weight="balanced",
    n_jobs=-1
)

rf_model.fit(X_train, y_train)

In [ ]:
# make predictions 
y_pred = rf_model.predict(X_test)

y_prob = rf_model.predict_proba(X_test)[:, 1]

In [ ]:
# metrics
# recall is important for scam transactions
# eg. if Recall = 0.83 (apprx caught 83% of scam transactions)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))

print()
print(classification_report(
    y_test,
    y_pred,
    target_names=["Legal", "Scam"]
))

# TN = legit transaction correctly found 
# FP = legit transaction wrongly flagged as scam 
# FN = scam transaction wrongly found as legit 
# TP = scam transaction correctly found 

In [ ]:
# Confusion Matrix 
cm = confusion_matrix(y_test, y_pred)

disp = ConfusionMatrixDisplay(
    confusion_matrix=cm,
    display_labels=["Legal", "Scam"]
)

disp.plot()

plt.title("Random Forest Transaction Classification")
plt.show()

# ROC Curve 

In [ ]:
fpr, tpr, thresholds = roc_curve(
    y_test,
    y_prob
)

auc = roc_auc_score(
    y_test,
    y_prob
)

plt.figure(figsize=(8, 6))

plt.plot(
    fpr,
    tpr,
    label=f"Random Forest AUC = {auc:.3f}"
)

plt.plot(
    [0, 1],
    [0, 1],
    linestyle="--",
    label="Random classifier"
)

plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")

plt.title("ROC Curve - Scam Transaction Detection")

plt.legend()

plt.show()

# feuture importance
# Which characs of Bitcoin transactions and address histories are most useful for detecting malicious activity? 
# maybe structure them as: 
# transaction = what's happening right now 
# input history = behaviour of wallets sending BTC
# output history = behaviour of wallets receiving BTC 
# relationship history = has sender/ receiver interacted before?

In [ ]:
feature_importance = pd.DataFrame({
    "feature": feature_columns,
    "importance": rf_model.feature_importances_
})

feature_importance = feature_importance.sort_values(
    by="importance",
    ascending=False
)

display(feature_importance)

top_features = feature_importance.head(15)

plt.figure(figsize=(10, 7))

plt.barh(
    top_features["feature"],
    top_features["importance"]
)

plt.xlabel("Random Forest Feature Importance")
plt.ylabel("Feature")

plt.title("Top 15 Features for Illicit Transaction Detection")

plt.gca().invert_yaxis()

plt.show()
